File: 4-Docker-Compose - Atomic Code/c/compose_netparser.c
/*
 * Lenguaje:   C (C11)
 * Asignatura: Lenguajes y compiladores
 * Equipo:     Atomic Code
 *             Segmentando el codigo, construyendo la logica
 *
 * Integrantes:
 *   - Victor Vargas    C.I: 30.697.219
 *   - Keibel Guilarte  C.I: 28.726.605
 *   - Oriana Marquez   C.I: 31.354.299
 *   - Jeanny Monagas   C.I: 30.857.471
 *
 * ---------------------------------------------------------------------------
 *                           DESCRIPCION
 * ---------------------------------------------------------------------------
 * Version en C del lexer + parser de interfaces de red de Docker Compose.
 * Mismo algoritmo y mismo formato de salida que python/compose_netparser.py
 * y rust/src/main.rs, para que el experimento de tiempos compare directamente
 * la implementacion del mismo analisis en tres lenguajes distintos.
 *
 * Lexico por linea: se ignoran lineas en blanco y comentarios ('#'). Cada
 * linea se clasifica en indentacion efectiva, clave y valor (o solo valor si
 * es un escalar suelto de una lista, ej. '- frontend').
 *
 * Sintactico: se mantiene una pila de (indentacion, clave) que reconstruye la
 * ruta de claves ancestras de cada linea, igual que en la version Python;
 * contra esa ruta se contrastan los mismos patrones fijos de
 * 'services.<svc>.networks' y 'networks.<red>...'.
 *
 * Uso:      compose_netparser.exe archivo.yml
 * ---------------------------------------------------------------------------
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#define MAX_LINEA 1024
#define MAX_CLAVE 128
#define MAX_PILA 32
#define MAX_REDES 128
#define MAX_INTERFACES 4096

typedef struct {
    char nombre[MAX_CLAVE];
    char driver[MAX_CLAVE];
    char subnet[MAX_CLAVE];
    char gateway[MAX_CLAVE];
} Red;

typedef struct {
    char servicio[MAX_CLAVE];
    char red[MAX_CLAVE];
    char ip[MAX_CLAVE];
} Interfaz;

typedef struct {
    int indent;
    char clave[MAX_CLAVE];
} NivelPila;

static Red redes[MAX_REDES];
static int total_redes = 0;

static Interfaz interfaces[MAX_INTERFACES];
static int total_interfaces = 0;

static NivelPila pila[MAX_PILA];
static int tope_pila = 0; /* pila[0] = raiz virtual, indent = -1 */

static int buscar_red(const char *nombre) {
    for (int i = 0; i < total_redes; i++) {
        if (strcmp(redes[i].nombre, nombre) == 0) return i;
    }
    return -1;
}

/* Copia 'origen' a 'destino' (tamano MAX_CLAVE) garantizando terminador nulo,
 * incluso si 'origen' es mas largo que el buffer (se trunca sin desbordar). */
static void copiar_clave(char *destino, const char *origen) {
    snprintf(destino, MAX_CLAVE, "%s", origen);
}

static void set_si_vacio(char *campo, const char *valor) {
    if (valor != NULL) {
        copiar_clave(campo, valor);
    }
}

/* Recorta espacios/CR/LF al final y espacios al inicio; devuelve puntero al
 * primer caracter no-espacio dentro del propio buffer (in place). */
static char *recortar(char *s) {
    size_t len = strlen(s);
    while (len > 0 && (s[len - 1] == '\n' || s[len - 1] == '\r' || s[len - 1] == ' ' || s[len - 1] == '\t')) {
        s[--len] = '\0';
    }
    char *inicio = s;
    while (*inicio == ' ' || *inicio == '\t') inicio++;
    return inicio;
}

/* Analiza una linea cruda. Devuelve 1 si produjo (indent, clave?, valor),
 * 0 si la linea debe ignorarse (vacia o comentario). */
static int analizar_linea(char *linea_cruda, int *indent_out, char clave_out[MAX_CLAVE],
                           int *tiene_clave, char valor_out[MAX_CLAVE], int *tiene_valor) {
    int indent = 0;
    char *p = linea_cruda;
    while (*p == ' ') { indent++; p++; }

    char *contenido = recortar(p);
    if (contenido[0] == '\0' || contenido[0] == '#') return 0;

    if (contenido[0] == '-' && contenido[1] == ' ') {
        contenido += 2;
        while (*contenido == ' ') contenido++;
        indent += 2;
    } else if (contenido[0] == '-' && contenido[1] == '\0') {
        return 0;
    }

    char *dos_puntos = strchr(contenido, ':');
    *indent_out = indent;
    *tiene_clave = 0;
    *tiene_valor = 0;
    clave_out[0] = '\0';
    valor_out[0] = '\0';

    if (dos_puntos != NULL) {
        size_t len_clave = (size_t)(dos_puntos - contenido);
        if (len_clave >= MAX_CLAVE) len_clave = MAX_CLAVE - 1;
        strncpy(clave_out, contenido, len_clave);
        clave_out[len_clave] = '\0';
        char *fin_clave = clave_out + strlen(clave_out);
        while (fin_clave > clave_out && fin_clave[-1] == ' ') *(--fin_clave) = '\0';
        *tiene_clave = 1;

        char *resto = dos_puntos + 1;
        while (*resto == ' ') resto++;
        char *valor_recortado = recortar(resto);
        if (valor_recortado[0] != '\0') {
            copiar_clave(valor_out, valor_recortado);
            *tiene_valor = 1;
        }
    } else {
        copiar_clave(valor_out, contenido);
        *tiene_valor = 1;
    }
    return 1;
}

static void procesar_linea(int indent, const char *clave, int tiene_clave, const char *valor, int tiene_valor) {
    while (tope_pila > 0 && pila[tope_pila - 1].indent >= indent) {
        tope_pila--;
    }

    /* ruta = claves de pila[1..tope_pila-1] + (clave o "*") */
    const char *ruta[MAX_PILA];
    int n = 0;
    for (int i = 1; i < tope_pila; i++) ruta[n++] = pila[i].clave;
    ruta[n++] = tiene_clave ? clave : "*";

    if (n == 2 && strcmp(ruta[0], "networks") == 0 && tiene_clave) {
        if (total_redes < MAX_REDES) {
            Red *r = &redes[total_redes++];
            copiar_clave(r->nombre, ruta[1]);
            copiar_clave(r->driver, "bridge");
            r->subnet[0] = '\0';
            r->gateway[0] = '\0';
        }
    } else if (n == 3 && strcmp(ruta[0], "networks") == 0 && strcmp(ruta[2], "driver") == 0) {
        int idx = buscar_red(ruta[1]);
        if (idx >= 0 && tiene_valor) set_si_vacio(redes[idx].driver, valor);
    } else if (n == 5 && strcmp(ruta[0], "networks") == 0 && strcmp(ruta[2], "ipam") == 0 &&
               strcmp(ruta[3], "config") == 0 && (strcmp(ruta[4], "subnet") == 0 || strcmp(ruta[4], "gateway") == 0)) {
        int idx = buscar_red(ruta[1]);
        if (idx >= 0 && tiene_valor) {
            if (strcmp(ruta[4], "subnet") == 0) set_si_vacio(redes[idx].subnet, valor);
            else set_si_vacio(redes[idx].gateway, valor);
        }
    } else if (n == 4 && strcmp(ruta[0], "services") == 0 && strcmp(ruta[2], "networks") == 0 &&
               strcmp(ruta[3], "*") == 0) {
        if (total_interfaces < MAX_INTERFACES && tiene_valor) {
            Interfaz *iface = &interfaces[total_interfaces++];
            copiar_clave(iface->servicio, ruta[1]);
            copiar_clave(iface->red, valor);
            iface->ip[0] = '\0';
        }
    } else if (n == 4 && strcmp(ruta[0], "services") == 0 && strcmp(ruta[2], "networks") == 0) {
        if (total_interfaces < MAX_INTERFACES) {
            Interfaz *iface = &interfaces[total_interfaces++];
            copiar_clave(iface->servicio, ruta[1]);
            copiar_clave(iface->red, ruta[3]);
            iface->ip[0] = '\0';
        }
    } else if (n == 5 && strcmp(ruta[0], "services") == 0 && strcmp(ruta[2], "networks") == 0 &&
               strcmp(ruta[4], "ipv4_address") == 0) {
        for (int i = total_interfaces - 1; i >= 0; i--) {
            if (strcmp(interfaces[i].servicio, ruta[1]) == 0 && strcmp(interfaces[i].red, ruta[3]) == 0) {
                if (tiene_valor) set_si_vacio(interfaces[i].ip, valor);
                break;
            }
        }
    }

    if (tiene_clave && tope_pila < MAX_PILA) {
        pila[tope_pila].indent = indent;
        copiar_clave(pila[tope_pila].clave, clave);
        tope_pila++;
    }
}

int main(int argc, char **argv) {
    if (argc != 2) {
        fprintf(stderr, "Uso: %s <archivo docker-compose.yml>\n", argv[0]);
        return 1;
    }
    FILE *archivo = fopen(argv[1], "r");
    if (!archivo) {
        perror("No se pudo abrir el archivo");
        return 1;
    }

    pila[0].indent = -1;
    pila[0].clave[0] = '\0';
    tope_pila = 1;

    char linea[MAX_LINEA];
    while (fgets(linea, sizeof(linea), archivo) != NULL) {
        int indent;
        char clave[MAX_CLAVE];
        char valor[MAX_CLAVE];
        int tiene_clave, tiene_valor;
        if (analizar_linea(linea, &indent, clave, &tiene_clave, valor, &tiene_valor)) {
            procesar_linea(indent, clave, tiene_clave, valor, tiene_valor);
        }
    }
    fclose(archivo);

    printf("REDES:\n");
    for (int i = 0; i < total_redes; i++) {
        printf("nombre=%s driver=%s subnet=%s gateway=%s\n",
               redes[i].nombre,
               redes[i].driver[0] ? redes[i].driver : "-",
               redes[i].subnet[0] ? redes[i].subnet : "-",
               redes[i].gateway[0] ? redes[i].gateway : "-");
    }
    printf("\nINTERFACES:\n");
    for (int i = 0; i < total_interfaces; i++) {
        printf("servicio=%s red=%s ip=%s\n",
               interfaces[i].servicio, interfaces[i].red,
               interfaces[i].ip[0] ? interfaces[i].ip : "-");
    }
    return 0;
}

File: 4-Docker-Compose - Atomic Code/python/compose_netparser.py
# Lenguaje:   Python 3
# Asignatura: Lenguajes y compiladores
# Equipo:     Atomic Code
#             Segmentando el código, construyendo la lógica
#
# Integrantes:
#   - Victor Vargas    C.I: 30.697.219
#   - Keibel Guilarte  C.I: 28.726.605
#   - Oriana Márquez   C.I: 31.354.299
#   - Jeanny Monagas   C.I: 30.857.471
#
# ---------------------------------------------------------------------------
#                           DESCRIPCIÓN
# ---------------------------------------------------------------------------
# Lexer + parser para extraer las interfaces de red de un archivo Docker
# Compose. NO es un parser YAML general: reconoce el subconjunto necesario
# para esta tarea (bloques 'services:' y 'networks:', con indentación
# uniforme de 2 espacios por nivel, que es como se generan las muestras de
# 'generar_muestras.py'). Esta misma versión (algoritmo y formato de salida)
# se reimplementa en C y Rust para el experimento de tiempos de ejecución,
# de modo que la comparación entre lenguajes sea justa.
#
# Léxico (por línea): se ignoran líneas en blanco y comentarios ('#'). Cada
# línea se clasifica en: indentación (espacios al inicio), si es ITEM de
# lista ('- '), y CLAVE: VALOR (o solo VALOR si es un escalar suelto de una
# lista, ej. '- frontend').
#
# Sintáctico: se mantiene una pila de (indentación, clave) que reconstruye la
# RUTA de claves ancestras de cada línea (como en un YAML real). Contra esa
# ruta se contrastan patrones fijos, ej. ["networks", <red>, "driver"], que
# es la "gramática" de lo que sabemos interpretar; toda ruta que no matchea
# ninguno de esos patrones (version, volumes, image, etc.) simplemente se
# ignora.
#
# Uso:      python compose_netparser.py archivo.yml
# Salida:   bloque REDES (una por línea) y bloque INTERFACES (una fila por
#           cada red a la que se conecta cada servicio), en un formato de
#           texto plano fijo e idéntico entre las 3 implementaciones.
# ---------------------------------------------------------------------------

import sys


def analizar_linea(linea_cruda):
    """Léxico de una línea: (indentacion_efectiva, clave_o_None, valor)."""
    sin_salto = linea_cruda.rstrip("\n").rstrip("\r")
    indent = len(sin_salto) - len(sin_salto.lstrip(" "))
    contenido = sin_salto.strip()
    if contenido == "" or contenido.startswith("#"):
        return None

    if contenido.startswith("- "):
        contenido = contenido[2:]
        indent += 2
    elif contenido == "-":
        return None

    if ":" in contenido:
        clave, _, valor = contenido.partition(":")
        clave = clave.strip()
        valor = valor.strip()
        if valor == "":
            valor = None
        return (indent, clave, valor)
    return (indent, None, contenido.strip())


def analizar_compose(lineas):
    """Recorre el archivo y devuelve (redes, interfaces).
    redes:      lista de dicts {nombre, driver, subnet, gateway}
    interfaces: lista de dicts {servicio, red, ip}
    """
    redes = []
    redes_por_nombre = {}
    interfaces = []

    pila = [(-1, None)]  # (indentacion, clave) ; raiz virtual

    for linea_cruda in lineas:
        analizado = analizar_linea(linea_cruda)
        if analizado is None:
            continue
        indent, clave, valor = analizado

        while pila[-1][0] >= indent:
            pila.pop()

        ruta = [k for _, k in pila[1:]]
        if clave is not None:
            ruta_completa = ruta + [clave]
        else:
            ruta_completa = ruta + ["*"]

        # --- reglas de la gramática que reconocemos ------------------------
        if ruta_completa == ["networks", clave] and clave is not None and len(ruta) == 1:
            redes_por_nombre[clave] = {"nombre": clave, "driver": "bridge", "subnet": None, "gateway": None}
            redes.append(redes_por_nombre[clave])
        elif len(ruta_completa) == 3 and ruta_completa[0] == "networks" and ruta_completa[2] == "driver":
            nombre_red = ruta_completa[1]
            if nombre_red in redes_por_nombre:
                redes_por_nombre[nombre_red]["driver"] = valor
        elif len(ruta_completa) == 5 and ruta_completa[0] == "networks" and ruta_completa[2] == "ipam" \
                and ruta_completa[3] == "config" and ruta_completa[4] in ("subnet", "gateway"):
            nombre_red = ruta_completa[1]
            if nombre_red in redes_por_nombre:
                redes_por_nombre[nombre_red][ruta_completa[4]] = valor
        elif len(ruta_completa) == 4 and ruta_completa[0] == "services" and ruta_completa[2] == "networks" \
                and ruta_completa[3] == "*":
            servicio = ruta_completa[1]
            interfaces.append({"servicio": servicio, "red": valor, "ip": None})
        elif len(ruta_completa) == 4 and ruta_completa[0] == "services" and ruta_completa[2] == "networks":
            servicio = ruta_completa[1]
            nombre_red = ruta_completa[3]
            interfaces.append({"servicio": servicio, "red": nombre_red, "ip": None})
        elif len(ruta_completa) == 5 and ruta_completa[0] == "services" and ruta_completa[2] == "networks" \
                and ruta_completa[4] == "ipv4_address":
            servicio = ruta_completa[1]
            nombre_red = ruta_completa[3]
            for interfaz in reversed(interfaces):
                if interfaz["servicio"] == servicio and interfaz["red"] == nombre_red:
                    interfaz["ip"] = valor
                    break

        if clave is not None:
            pila.append((indent, clave))

    return redes, interfaces


def formatear_salida(redes, interfaces):
    lineas = ["REDES:"]
    for red in redes:
        lineas.append(
            f"nombre={red['nombre']} driver={red['driver'] or '-'} "
            f"subnet={red['subnet'] or '-'} gateway={red['gateway'] or '-'}"
        )
    lineas.append("")
    lineas.append("INTERFACES:")
    for interfaz in interfaces:
        lineas.append(
            f"servicio={interfaz['servicio']} red={interfaz['red']} ip={interfaz['ip'] or '-'}"
        )
    return "\n".join(lineas)


def main():
    if len(sys.argv) != 2:
        print("Uso: python compose_netparser.py <archivo docker-compose.yml>", file=sys.stderr)
        sys.exit(1)
    with open(sys.argv[1], "r", encoding="utf-8") as archivo:
        lineas = archivo.readlines()
    redes, interfaces = analizar_compose(lineas)
    print(formatear_salida(redes, interfaces))


if __name__ == "__main__":
    main()
